File: cpp/7-ReverseInteger/main.cpp
#include <cstdlib>
#include <iostream>

using namespace std;

class Solution 
{
public:
    int reverse(int x) 
    {
        const int maxInt = 2147483647;

        const auto isNegative = x < 0;
        int value = 0;

        while(x)
        {
            const auto nextDigit = abs(x % 10);

            // Check for overflow
            if(value && ((maxInt - nextDigit) / value) < 10) return 0;
            value = (value * 10) + nextDigit;

            x /= 10;
        }

        // A positive integer always has a negative equivilant
        if(isNegative) value *= -1;

        return value;
    }
};

int main()
{
    Solution solution;
    std::cout << solution.reverse(123) << "\n";
    std::cout << solution.reverse(-123) << "\n";
    std::cout << solution.reverse(120) << "\n";
    std::cout << solution.reverse(0) << "\n";
}

File: rust/7-ReverseInteger/main.rs
pub fn reverse(x: i32) -> i32 {
    let max_int : i32 = 2147483647;
    let is_negative = (x < 0);
    let mut value = 0;

    let mut x = x;
    while x != 0 {
        let next_digit = (x % 10).abs();

        if value != 0 && ((max_int - next_digit) / value) < 10 {
            return 0;
        }

        value = (value * 10) + next_digit;
        x /= 10;
    }

    if is_negative {
        value *= -1;
    }

    value
}

fn main() 
{
    println!("{:?}",reverse(103));
    println!("{:?}",reverse(-987));
}